File: 10. Regular Expression Matching Hard/solution.go
package main

import "fmt"

func main() {
	fmt.Println(isMatch("aa", "a"))
	fmt.Println(isMatch("aa", "a*"))
	fmt.Println(isMatch("ab", ".*"))
}

func isMatch(s string, p string) bool {
	memo := make(map[string]bool)

	var dp func(int, int) bool
	dp = func(i, j int) bool {
		key := fmt.Sprintf("%d,%d", i, j)
		if val, ok := memo[key]; ok {
			return val
		}

		var ans bool
		if j == len(p) {
			ans = i == len(s)
		} else {
			firstMatch := i < len(s) && (p[j] == s[i] || p[j] == '.')

			if j+1 < len(p) && p[j+1] == '*' {
				ans = dp(i, j+2) || (firstMatch && dp(i+1, j))
			} else {
				ans = firstMatch && dp(i+1, j+1)
			}
		}

		memo[key] = ans
		return ans
	}

	return dp(0, 0)
}

File: 10. Regular Expression Matching Hard/solution.py
def is_match(s: str, p: str) -> bool:
    memo = {}
    
    def dp(i: int, j: int) -> bool:
        if (i, j) in memo:
            return memo[(i, j)]

        if j == len(p):
            ans = i == len(s)
        else:
            first_match = i < len(s) and p[j] in {s[i], '.'}

            if j + 1 < len(p) and p[j + 1] == '*':
                ans = dp(i, j + 2) or (first_match and dp(i + 1, j))
            else:
                ans = first_match and dp(i + 1, j + 1)

        memo[(i, j)] = ans
        return ans

    return dp(0, 0)
    

print(is_match('aa', 'a'))
print(is_match('aa', 'a*'))
print(is_match('ab', '.*'))